[GStreamer][WebRTC] Disabled incoming audio track breaks webaudio provider

philn opened the pull request:
#### 5e1b9418e1ae6ba8954e7f6a916a7882e03ac69e
<pre>
[GStreamer][WebRTC] Disabled incoming audio track breaks webaudio provider
<a href="https://bugs.webkit.org/show_bug.cgi?id=252751">https://bugs.webkit.org/show_bug.cgi?id=252751</a>

Reviewed by Xabier Rodriguez-Calvar.

Disabling an incoming audio track internally means that the mediastreamsrc switches from RTP
payloads to raw audio. The AudioSourceProvider wasn&apos;t able to reliably handle this, but switching
its internal pipeline to uridecodebin3 solves the issue. Internally uridecodebin3 will be able to
reconfigure properly.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/audio/gstreamer/AudioSourceProviderGStreamer.cpp:
(WebCore::AudioSourceProviderGStreamer::AudioSourceProviderGStreamer):
(WebCore::AudioSourceProviderGStreamer::~AudioSourceProviderGStreamer):
(WebCore::AudioSourceProviderGStreamer::setClient):
* Source/WebCore/platform/audio/gstreamer/AudioSourceProviderGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/260742@main">https://commits.webkit.org/260742@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1778c76a8e1445e14ed961e8a0bd7a05a742d34c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109214 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18298 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42027 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/741 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118444 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113096 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19757 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9599 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101457 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114970 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11071 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11797 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17166 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50647 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7412 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13420 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/audio/gstreamer/AudioSourceProviderGStreamer.cpp:
```diff
@@ -90,18 +91,33 @@ AudioSourceProviderGStreamer::AudioSourceProviderGStreamer()
+
+    g_signal_connect_swapped(decodebin, "source-setup", G_CALLBACK(+[](AudioSourceProviderGStreamer* provider, GstElement* sourceElement) {
+        ASSERT(WEBKIT_IS_MEDIA_STREAM_SRC(sourceElement));
+        if (!WEBKIT_IS_MEDIA_STREAM_SRC(sourceElement))
```
Instead of `ASSERT` and then checking and bailing out, you can just `ASSERT_NOT_REACHED` inside the if.

philn committed:
[GStreamer][WebRTC] Disabled incoming audio track breaks webaudio provider
https://bugs.webkit.org/show_bug.cgi?id=252751

Reviewed by Xabier Rodriguez-Calvar.

Disabling an incoming audio track internally means that the mediastreamsrc switches from RTP
payloads to raw audio. The AudioSourceProvider wasn't able to reliably handle this, but switching
its internal pipeline to uridecodebin3 solves the issue. Internally uridecodebin3 will be able to
reconfigure properly.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/audio/gstreamer/AudioSourceProviderGStreamer.cpp:
(WebCore::AudioSourceProviderGStreamer::AudioSourceProviderGStreamer):
(WebCore::AudioSourceProviderGStreamer::~AudioSourceProviderGStreamer):
(WebCore::AudioSourceProviderGStreamer::setClient):
* Source/WebCore/platform/audio/gstreamer/AudioSourceProviderGStreamer.h:

Canonical link: https://commits.webkit.org/260742@main
